diffrax/_solver/kl.py

```diff
@@ -0,0 +1,291 @@
+    """
+    drift1 = drift_term1.vf(t0, y0, args)
+    drift2 = drift_term2.vf(t0, y0, args)
+    drift = jtu.tree_map(operator.sub, drift1, drift2)
+
+    diffusion = diffusion_term.vf(t0, y0, args)  # assumes same diffusion
+
+    divergences = lx.linear_solve(diffusion, drift, solver=linear_solver).value
+
+    kl_divergence = jtu.tree_map(lambda x: 0.5 * jnp.sum(x**2), divergences)
+    kl_divergence = jtu.tree_reduce(operator.add, kl_divergence)
+
+    return KLState(drift1, jnp.squeeze(kl_divergence))
+
+
+class _KLDrift(AbstractTerm):
+    drift1: ODETerm
+    drift2: ODETerm
+    diffusion: ControlTerm
+    linear_solver: lx.AbstractLinearSolver
+
+    def vf(self, t: RealScalarLike, y: Y, args: Args) -> KLState:
+        y = y.y
+        return _compute_kl_integral(
+            self.drift1, self.drift2, self.diffusion, t, y, args, self.linear_solver
+        )
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> Control:
+        return t1 - t0
+
+    def prod(self, vf: VF, control: RealScalarLike) -> Y:
+        return jtu.tree_map(lambda v: control * v, vf)
+
+
+class _KLControlTerm(AbstractTerm):
+    control_term: ControlTerm
+
+    def vf(self, t: RealScalarLike, y: Y, args: Args) -> KLState:
+        y = y.y
+        vf = self.control_term.vf(t, y, args)
+        return KLState(vf, jnp.array(0.0))
+
+    def contr(self, t0: RealScalarLike, t1: RealScalarLike, **kwargs) -> KLState:
+        return KLState(self.control_term.contr(t0, t1), jnp.array(0.0))
+
+    def vf_prod(self, t: RealScalarLike, y: Y, args: Args, control: Control) -> KLState:
+        y = y.y
+        control = control.y
+        return KLState(self.control_term.vf_prod(t, y, args, control), jnp.array(0.0))
+
+    def prod(self, vf: VF, control: Control) -> KLState:
+        vf = vf.y
+        control = control.y
+        return KLState(self.control_term.prod(vf, control), jnp.array(0.0))
+
+
+class KLSolver(AbstractWrappedSolver[_SolverState]):
+    r"""Given an SDE of the form
+
+    $$
+    \mathrm{d}y(t) = f_\theta (t, y(t)) dt + g_\phi (t, y(t)) dW(t) \qquad \zeta_\theta (ts[0]) = y_0
+    $$
+
+    $$
+    \mathrm{d}z(t) = h_\psi (t, z(t)) dt + g_\phi (t, z(t)) dW(t) \qquad \nu_\psi (ts[0]) = z_0
+    $$
+
+    compute:
+
+    $$
+    \int_{ts[i-1]}^{ts[i]} g_\phi (t, y(t))^{-1} (f_\theta (t, y(y)) - h_\psi (t, y(t))) dt
+    $$
+
+    for every time interval. This is useful for KL based latent SDEs. The output
+    of the solution.ys will be a tuple containing (ys, kls) where kls is the KL
+    divergence integration at that time. Unless the noise is diagonal, this
+    inverse can be extremely costly for higher dimenions.
+
+    The input must be a `MultiTerm` composed of the first SDE with drift `f`
+    and diffusion `g` and the second either a SDE or just the drift term
+    (since the diffusion is assumed to be the same). For example, a type
+    of: `MuliTerm(MultiTerm(ODETerm, _DiffusionTerm), ODETerm)`.
```

Comment:
As per this comment:
https://github.com/patrick-kidger/diffrax/pull/402#pullrequestreview-2013511154
and also the updated term docs:
https://docs.kidger.site/diffrax/api/terms/
then this outer `MultiTerm` isn't really in-keeping. We're not adding all of these extra terms on to the same evolving state.

Bearing in mind that the rest of Diffrax has to see this as just another SDE solve.

I think this one might take a bit more iteration to get to something that's obeying the abstractions in the way they're designed, I'm afraid.

Reply:
I think I get what you're saying, multiterm implies a single differential equation "unit". So the composed multi terms is bad form. However, I'm not sure I see the difficulty going forward, I can replace it with tuple (multiterm, multiterm) or even tuple (multiterm, ode term). Which seems to adhere to this principle of multiterm = sde unit, since we are integrating two simultaneous SDEs, while also falling in line with other solvers (such as implicit Euler as you remarked). 

On the terms vs solver approach, I am open to both. I think in my many iterations/experimentations I found the solver approach more in line with my thinking about the nature of the problem, specifically the original idea of (terms, kl_term) I didn't see as appealing since the KL_term relies on information from the other term and I didn't see a clean way to do that. However, having terms with a term wrapper is very doable (but may not mesh with the repo as well).

Reply:
Maybe introducing a totally new term is ok (given the remarks in https://github.com/patrick-kidger/diffrax/pull/453), in which case the approach of a KLTerm (rather than a solver), is doable. Given the restricted nature of terms so far, I originally thought that wasn't in line with the package